### DISCUSSION TITLE
GPUでジョブを実行する際のスレッド数の指定について

### DISCUSSION THREAD

**Discussion by Yonaka-no-12ji** [nig-sc/nigsc_homepage2#295] [Q&A] [answered]
scRNAseq解析関連でご助言をいただきたく質問させていただきました。

現在、遺伝研スパコンでscvitoolsを遺伝研スパコンで利用し、singlecellデータの解析を行わせていただいています。
GPUでジョブを実行しているのですが、スレッド数が足りない旨のエラーが毎回発生します。
調べてみるとCUDAでスレッド数を指定できるようなのですが、遺伝研スパコンのGPUでスレッド数を指定してジョブを行うにはどのようにすればよいか分からず悩んでおります。ご助言いただけますと幸いです。

**Comment by sc-fs** [comment, marked as answer]
状況の確認のため、
実際に発生してるエラー及び実行したジョブ番号を教えていただけないでしょうか。

**Reply by Yonaka-no-12ji** [reply]
返信ありがとうございます。
ジョブ番号は23614013です。ジョブの内容はscvitoolsのtutrialで紹介されているデータの前処理です。
エラーメッセージは以下のものです。
F external/org_tensorflow/tensorflow/tsl/platform/default/env.cc:74] Check failed: ret == 0 (11 vs. 0)Thread tf_pjrt_thread_pool creation via pthread_create() failed.
よろしくお願いします。

**Reply by sc-fs** [reply]
スレッド数の制限に達したと言うエラーになるようですので、起動プログラム側で制御できないか等、こちらでも、同様なスクリプトを準備して、調査します。

**Reply by sc-fs** [reply]
確実な原因が判明していないのですが、
GPU ドライバーのバージョンと CUDA のバージョンのずれで発生する可能性があると言うことなので、
singularityイメージでは無く、
以下を参考にcondaを利用した仮想環境上に環境を準備して実行して頂けませんか。
https://docs.scvi-tools.org/en/stable/installation.html

あらかじめcondaの環境は準備してください。

こちらで実行した手順（gpuのログインノードに実施。）
`$ qlogin -l gpu -l cuda=1 -l mem_req=24G,s_vmem=24G

必要であれば、condaの環境設定
$ export PATH=$HOME/anaconda3/bin:$PATH;
$ source ~/anaconda3/etc/profile.d/conda.sh

以下の上記webページの内容
$ conda create -n scvi-env python=3.9
$conda activate scvi-env

cudaが現在11.6なので、11.6を指定
(scvi-env) $ conda install pytorch torchvision torchaudio pytorch-cuda=11.6 -c pytorch -c nvidia
(scvi-env) $ conda install jax jaxlib -c conda-forge
(scvi-env) $ conda install scvi-tools -c conda-forge
(scvi-env) $ conda install -c conda-forge r-base r-essentials r-reticulate

(scvi-env) $ pip install scvi-tools
(scvi-env) $ pip install "scvi-tools[tutorials]"


実行するとバージョンの違いでエラーが発生したので、再インストールを実施
(scvi-env) $ pip install "numpy<1.21.0"
(scvi-env) $ pip install --upgrade pip
(scvi-env) $ pip install "jax[cuda11_cudnn82]" -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html

tutrialの実行
$ python scvitools_tutrial.py
一部warningが発生しますが、一応正常処理されます。

MuData object with n_obs × n_vars = 7947 × 14309
  2 modalities
    rna:        7947 x 14309
      obs:      'batch', 'n_counts'
      var:      'gene_ids-0', 'feature_types-0', 'genome-0', 'gene_ids-1', 'n_counts'
      uns:      'log1p'
    log_norm_rna:       7947 x 14309
      obs:      'batch', 'n_counts'
      var:      'gene_ids-0', 'feature_types-0', 'genome-0', 'gene_ids-1', 'n_counts'
      uns:      'log1p'
INFO     Using column names from columns of adata.obsm['prot']
INFO     Computing empirical prior initialization for protein background.
Anndata setup with scvi-tools version 0.19.0.

Setup via `TOTALVI.setup_anndata` with arguments:
{
x   'protein_expression_obsm_key': 'prot',
x   'protein_names_uns_key': None,
x   'batch_key': None,
x   'layer': None,
x   'size_factor_key': None,
x   'categorical_covariate_keys': None,
x   'continuous_covariate_keys': None
}

         Summary Statistics
lqqqqqqqqqqqqqqqqqqqqqqqqqqwqqqqqqqk
x     Summary Stat Key     x Value x
tqqqqqqqqqqqqqqqqqqqqqqqqqqnqqqqqqqu
x         n_batch          x   1   x
x         n_cells          x 5247  x
x n_extra_categorical_covs x   0   x
x n_extra_continuous_covs  x   0   x
x         n_labels         x   1   x
x        n_proteins        x  32   x
x          n_vars          x 33538 x
mqqqqqqqqqqqqqqqqqqqqqqqqqqvqqqqqqqj
               Data Registry
lqqqqqqqqqqqqqqwqqqqqqqqqqqqqqqqqqqqqqqqqqqk
x Registry Key x    scvi-tools Location    x
tqqqqqqqqqqqqqqnqqqqqqqqqqqqqqqqqqqqqqqqqqqu
x      X       x          adata.X          x
x    batch     x adata.obs['_scvi_batch']  x
x    labels    x adata.obs['_scvi_labels'] x
x   proteins   x    adata.obsm['prot']     x
mqqqqqqqqqqqqqqvqqqqqqqqqqqqqqqqqqqqqqqqqqqj
                     labels State Registry
lqqqqqqqqqqqqqqqqqqqqqqqqqqqwqqqqqqqqqqqqwqqqqqqqqqqqqqqqqqqqqqk
x      Source Location      x Categories x scvi-tools Encoding x
tqqqqqqqqqqqqqqqqqqqqqqqqqqqnqqqqqqqqqqqqnqqqqqqqqqqqqqqqqqqqqqu
x adata.obs['_scvi_labels'] x     0      x          0          x
mqqqqqqqqqqqqqqqqqqqqqqqqqqqvqqqqqqqqqqqqvqqqqqqqqqqqqqqqqqqqqqj
                     batch State Registry
lqqqqqqqqqqqqqqqqqqqqqqqqqqwqqqqqqqqqqqqwqqqqqqqqqqqqqqqqqqqqqk
x     Source Location      x Categories x scvi-tools Encoding x
tqqqqqqqqqqqqqqqqqqqqqqqqqqnqqqqqqqqqqqqnqqqqqqqqqqqqqqqqqqqqqu
x adata.obs['_scvi_batch'] x     0      x          0          x
mqqqqqqqqqqqqqqqqqqqqqqqqqqvqqqqqqqqqqqqvqqqqqqqqqqqqqqqqqqqqqj
$

tutrialファイルの内容
$cat scvitools_tutrial.py
import muon
import numpy as np
import scanpy as sc
import scvi

#loading pbmc3k
pbmc3k = scvi.data.read_h5ad("pbmc3k_raw.h5ad")
print("pbmc3k Data")
pbmc3k
pbmc3k.var.head()

#loading pbmc5k
pbmc5k = muon.read_10x_h5(
    "5k_pbmc_protein_v3_filtered_feature_bc_matrix.h5",
)
print("pbmc5k Data")
pbmc5k
pbmc5k.var_names_make_unique()

#connect pbmc3k and pbmc5k
adata = pbmc5k.mod["rna"].concatenate(pbmc3k)
print("check result")
adata.obs.sample(n=5)

#preprocessing the data
sc.pp.filter_genes(adata, min_counts=3)
sc.pp.filter_cells(adata, min_counts=3)
adata.layers["counts"] = adata.X.copy()
sc.pp.normalize_total(adata)
sc.pp.log1p(adata)
adata.raw = adata
mdata = muon.MuData({"rna": adata.copy(), "log_norm_rna": adata.copy()}, axis=-1)
# Now rna is count-based and log_norm_rna is log-normalized
mdata.mod["rna"].X = mdata.mod["rna"].layers["counts"]
del mdata.mod["rna"].raw
del mdata.mod["rna"].layers["counts"]
del mdata.mod["log_norm_rna"].layers["counts"]
print(mdata)

scvi.model.SCVI.setup_anndata(adata, layer="counts", batch_key="batch")

# totalVI requires dense protein data
pbmc5k.mod["prot"].X = np.asarray(pbmc5k.mod["prot"].X.A)
scvi.model.TOTALVI.setup_mudata(
    pbmc5k,
    protein_layer=None,
    rna_layer=None,
    modalities={"protein_layer": "prot", "rna_layer": "rna"},
)

adata_pbm5k = pbmc5k.mod["rna"]
adata_pbm5k.obsm["prot"] = pbmc5k.mod["prot"].to_df()

scvi.model.TOTALVI.setup_anndata(
    adata_pbm5k,
    protein_expression_obsm_key="prot",
)
model = scvi.model.TOTALVI(adata_pbm5k)
model.view_anndata_setup(adata_pbm5k)

`

**Reply by Yonaka-no-12ji** [reply]
ご助言いただきありがとうございます。試してみます。

**Comment by khigashi1987** [comment]
2022年末のPython講習会にご参加いただいた方でしょうか？以下は、講習会参加者の方から類似したご質問をいただいた際に回答した内容なのですが、ご参考にしていただけるかもしれません。

scVIは内部でpytorchの深層学習計算が動いており、pytorchをインストールするとき、pytorchが適切にマシンのGPUを認識するためにCUDAドライバと紐づける必要があります。
scVIの公式ページに掲載されているインストール手順は、CUDAドライバのバージョン “11.6” とリンクするための手順です。
ですが、遺伝研スパコンのGPUノードにはCUDAドライバ11.6が入っていないようです。
以下のようにmodule loadでドライバを切り替えられるようですが、11.6はありません。
https://sc.ddbj.nig.ac.jp/software/environmental_modules/
そして現状、pytorchはCUDAドライバの11.6, 11.7以外では簡単にインストールする手順が提供されていません。
したがって現状、遺伝研スパコンのGPUノードではpytorchによる深層学習は実行が難しい、という回答になります。

GPUを利用したscvi-toolsの計算を実行するためには、NVIDIAのGPUがささったLinuxかWindowsのマシンを用意して、新しいバージョンのCUDAドライバをインストールしていただくのが確実だと思います。

**Reply by Yonaka-no-12ji** [reply]
ご助言いただきありがとうございます。

**Reply by sc-fs** [reply]
遺伝研スパコンのCUDAドライババージョンですが、本日時点ではCUDA11.6となっています。
また、CUDAのバージョンアップは定期メンテナンスで基本的には実施します。
`[ashizawa@igt001 ~]$ nvidia-smi
Tue Feb 28 18:00:22 2023
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 510.47.03    Driver Version: 510.47.03    CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000000:16:00.0 Off |                    0 |
| N/A   36C    P0    40W / 300W |      3MiB / 16384MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  On   | 00000000:3A:00.0 Off |                    0 |
| N/A   35C    P0    40W / 300W |      3MiB / 16384MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
[ashizawa@igt001 ~]$`
